// @discord-nestjs/core/dist/discord-host.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DiscordHostModule = void 0;
const common_1 = require("@nestjs/common");
const discord_client_provider_alias_1 = require("./definitions/constants/discord-client-provider-alias");
const discord_command_provider_alias_1 = require("./definitions/constants/discord-command-provider-alias");
const discord_module_contant_1 = require("./definitions/constants/discord-module.contant");
const reflect_metadata_provider_alias_1 = require("./definitions/constants/reflect-metadata-provider-alias");
const discord_client_provider_1 = require("./providers/discord-client.provider");
const discord_command_provider_1 = require("./providers/discord-command.provider");
const reflect_metadata_provider_1 = require("./providers/reflect-metadata.provider");
const client_service_1 = require("./services/client.service");
const option_service_1 = require("./services/option.service");
let DiscordHostModule = class DiscordHostModule {
};
DiscordHostModule = __decorate([
    (0, common_1.Module)({
        providers: [
            client_service_1.ClientService,
            option_service_1.OptionService,
            {
                provide: reflect_metadata_provider_alias_1.REFLECT_METADATA_PROVIDER_ALIAS,
                useClass: reflect_metadata_provider_1.ReflectMetadataProvider,
            },
            {
                provide: discord_command_provider_alias_1.DISCORD_COMMAND_PROVIDER_ALIAS,
                useClass: discord_command_provider_1.DiscordCommandProvider,
            },
            {
                provide: discord_client_provider_alias_1.DISCORD_CLIENT_PROVIDER_ALIAS,
                useClass: discord_client_provider_1.DiscordClientProvider,
            },
            {
                provide: discord_module_contant_1.DISCORD_MODULE_OPTIONS,
                useFactory: () => ({}),
            },
        ],
        exports: [
            client_service_1.ClientService,
            option_service_1.OptionService,
            reflect_metadata_provider_alias_1.REFLECT_METADATA_PROVIDER_ALIAS,
            discord_command_provider_alias_1.DISCORD_COMMAND_PROVIDER_ALIAS,
            discord_client_provider_alias_1.DISCORD_CLIENT_PROVIDER_ALIAS,
            discord_module_contant_1.DISCORD_MODULE_OPTIONS,
        ],
    })
], DiscordHostModule);
exports.DiscordHostModule = DiscordHostModule;
